ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -133,107 +221,237 @@ long getMaxBlobSize() {
       }
+      metrics.statsOnDemandScanTimePerIndexSegmentMs.update(time.milliseconds() - indexSegmentStartProcessTimeMs,
+          TimeUnit.MILLISECONDS);
+    }
+    metrics.statsOnDemandScanTotalTimeMs.update(time.milliseconds() - startTimeMs, TimeUnit.MILLISECONDS);
+    if (validSizePerLogSegment.isEmpty()) {
+      validSizePerLogSegment.put(index.getStartOffset().getName(), 0L);
     }
     return validSizePerLogSegment;
   }
 
   /**
-   * Get a {@link List} of valid {@link IndexValue} in a given {@link IndexSegment}.
-   * @param indexSegment the indexSegment to be read from
-   * @param deleteReferenceTimeInMs the reference time in ms until which deletes are relevant
-   * @param expirationReferenceTimeInMs the reference time in ms until which expiration are relevant
-   * @param deletedKeys a {@link Set} of deleted keys to help determine whether a put is deleted
-   * @return a {@link List} of valid {@link IndexValue}
+   * Get all {@link IndexEntry} in a given {@link IndexSegment}.
+   * @param indexSegment the {@link IndexSegment} to fetch the index entries from
+   * @return a {@link List} of all {@link IndexEntry} in the given {@link IndexSegment}
    * @throws StoreException
    */
-  private List<IndexValue> getValidIndexValues(IndexSegment indexSegment, long deleteReferenceTimeInMs,
-      long expirationReferenceTimeInMs, Set<StoreKey> deletedKeys) throws StoreException {
+  private List<IndexEntry> getIndexEntries(IndexSegment indexSegment) throws StoreException {
     diskIOScheduler.getSlice(BlobStoreStats.IO_SCHEDULER_JOB_TYPE, BlobStoreStats.IO_SCHEDULER_JOB_ID, 1);
     List<IndexEntry> indexEntries = new ArrayList<>();
-    List<IndexValue> validIndexValues = new ArrayList<>();
     try {
       indexSegment.getIndexEntriesSince(null, new FindEntriesCondition(Integer.MAX_VALUE), indexEntries,
           new AtomicLong(0));
     } catch (IOException e) {
-      throw new StoreException("I/O error while getting entries from index segment", e, StoreErrorCodes.IOError);
+      throw new StoreException("I/O exception while getting entries from index segment", e, StoreErrorCodes.IOError);
     }
+    return indexEntries;
+  }
+
+  /**
+   * Get a {@link List} of valid {@link IndexValue}s from a given {@link List} of {@link IndexEntry} that belong to the
+   * same {@link IndexSegment}.
+   * @param indexSegment the {@link IndexSegment} where the entries came from
+   * @param indexEntries a {@link List} of unfiltered {@link IndexEntry} whose elements could be valid or invalid
+   * @param referenceTimeInMs the reference time in ms until which deletes and expiration are relevant
+   * @param deletedKeys a {@link Map} of deleted keys to operation time. Used to determine whether a PUT is deleted
+   * @return a {@link List} of valid {@link IndexValue}
+   * @throws StoreException
+   */
+  private List<IndexEntry> getValidIndexEntries(IndexSegment indexSegment, List<IndexEntry> indexEntries,
```

Comment:
this method not only gets valid index entries but also updated deletedKeys. Try to see if we can name it appropriately. For sometime, I was thinking this method just return valid index entries only.